### DISCUSSION TITLE
Can't depoy Palmer Penguins example on Shiny-server

### DISCUSSION THREAD

**Discussion by AlekseiBogachev** [quarto-dev/quarto-cli#10356] [Q&A] [answered]
### Description

Hi all!

I'm trying to deploy an Palmer Penguins example from the [Dashboards with Shiny for Python](https://quarto.org/docs/dashboards/interactivity/shiny-python/index.html) guide on Shiny Server.

I copied qmd-file, rendered it with Quarto, however Shiny Server threw an error:

```
shiny_server-1  | [2024-07-23T02:58:55.178] [INFO] Shiny Server - Error getting worker: Error: The application exited during initialization.
```

An uninformative message was recorded in the log:

```
Error: lexical error: invalid char in json text.
                                       NA
                     (right here) ------^
Execution halted
```

**Please, tell me how to deploy this example on Shiny Server. Is it possible??**

I use it in docker container. I tried to run [Restaurant tipping example](https://shiny.posit.co/py/) (got it with command `shiny create --template dashboard-tips`) and it works well.

And one more question bothers me: should I use Shiny-server for dashboards or is it enough to use Shiny for Python, that is, deploy the environment in a container and run the `shiny run app.py command`??

Thanks in advance for your help!

**Comment by cscheid** [comment]
It looks like you're asking a Shiny question. This is the Quarto forum. You'll probably get better luck on the Shiny forum.

**Reply by AlekseiBogachev** [reply]
OK, thank you!

**Reply by mcanouil** [reply]
@cscheid could it be a regression?

- https://github.com/quarto-dev/quarto-cli/issues/7347

PS: @AlekseiBogachev asking the same thing on multiple threads won't give you faster or better results/answers. You could have simply reference the closed issues.

What version of Quarto and shiny are you using? I would suggest you upgrade everything.

**Reply by AlekseiBogachev** [reply]
Hi! Thank you for your advice!

I just updated quarto and shiny-sever to the latest versions. Now I'm using **quarto-1.5.55-linux-amd64** , **shiny-server-1.5.22.1017-amd64.deb** and python packages  **shiny = "^0.10.2"**, **shinywidgets = "^0.3.2"**. However, the same error.

**Reply by cderv** [reply]
> I copied qmd-file, rendered it with Quarto, however Shiny Server threw an error:

How do you render ? What step are you following to deploy on shiny-server ? Which guide did you follow ? 

Having more information would help. Shiny interactive document with Quarto are shiny pre-rendered document and so must be deployed on Shiny server as such, like Rmd shiny prerendered document.

**Reply by AlekseiBogachev** [reply]
Hi!
Thanks for your answer!
> Having more information would help. Shiny interactive document with Quarto are shiny pre-rendered document and so must be deployed on Shiny server as such, like Rmd shiny prerendered document.

I created test project so you can reproduce my case.
https://github.com/AlekseiBogachev/penguins-dashboard-docker

When you build the docker image, an environment with all software and packages installed is created and the dashboard is rendered with `poetry run quarto render penguins.qmd` command.

When you run the image, the Shiny server starts on the port **3838**.
Logs are written in the `/dockeruser/penguins-dashboard/logs` directory inside the container.

You can use **build.sh** and **start.sh** to build and run the container.

**Reply by mcanouil** [reply]
As already answered, your issue is a shiny issue as the document itself works.
Please read the documentation on how to setup a shiny server, see <https://shiny.posit.co/py/docs/deploy-on-prem.html>.
Also, I suggest you try using the default setup and try the default app first to see if your setup is working.
And if you still have issue, please do ask on shiny related places which is not this repository.

**Reply by cderv** [reply, marked as answer]
First, @mcanouil yes this is shiny related, but Quarto does integrate with shiny and trying to make a shiny interactive document with Quarto is a valid Quarto question to me that is worth looking into. We even have documentation on our website regarding how to deploy shiny interactive document.  

----

Then, @AlekseiBogachev let me try to give you more context about what I was saying about shiny pre-rendered document. 
Thanks a lot for the Dockerfile by the way to help reproduce. Really helpful ! 

Looking at your dockerfile, I see you 
- `quarto render` the document 
- and then server the all folder using `shiny-server`, including the source `penguins.qmd`

In short, this is the presence of the source of this document which is causing problem on `shiny-server`. Just remove or rename the `penguins.qmd` in the hosted folder and your app will work. So in your dockerfile 

````diff
RUN poetry run quarto render penguins.qmd
+RUN rm penguins.qmd
````

Then your app will work. Let met explain why below

---- 

You are encountering some language specific difference of deploying shiny interactive application with Quarto.  Shiny for Python documents , and Shiny for R documents are not rendered and deployed the same. 

In the case of Shiny for Python document (which you are using here), see our documentation 
https://quarto.org/docs/dashboards/interactivity/shiny-python/running.html#deployment

> Your deployed application will consist of both the HTML document and app.py script generated by quarto render:
> ```
> example.html
> example_files/
> app.py
> ```

For shiny for python document, the .qmd source is not needed, and even can cause problem if present. 

This is because for Shiny for R, documents are shiny-prerendered document and `.qmd` source is required. Not all clear from our doc (https://quarto.org/docs/interactive/shiny/running.html#deployment) but still shows that process is different. 

Open Source shiny server is supporting Python application and you can find docs at: https://shiny.posit.co/py/docs/deploy-on-prem.html#deploy-to-shiny-server-open-source. It requires only `.html` file, resources folder, and `app.py`. 

It supports Quarto Shiny for R interactive document since recently (https://github.com/rstudio/shiny-server/pull/531) and for this it will treat any folder with a .qmd document as a shiny for R application, using shiny-prerendered HTML. 

This could be seen as an issue in shiny-server maybe - I don't know. 

Hope this helps understand how to connect all the docs together, and better comprehend the Quarto Shiny interactive document specificity.

**Reply by cderv** [reply]
I have opened an issue for track the problem in `shiny-server`. I believe this is known issue but it requires a new shiny-server version (which does not happen quite often) 
- https://github.com/rstudio/shiny-server/issues/574

For now, anyone having this problem needs to not deploy the .qmd when using a python app.

**Reply by AlekseiBogachev** [reply]
Thank you very much for the solution and the comprehensive explanation!
**Now the dashboard works!**
You have saved me a lot of time. You are incredible, thank you!
